cli/command/stack/list.go

```diff
@@ -18,18 +32,168 @@ func newListCommand(dockerCli command.Cli) *cobra.Command {
+	flags.SetAnnotation("namespace", "kubernetes", nil)
+	flags.SetAnnotation("namespace", "experimentalCLI", nil)
+	flags.BoolVarP(&opts.AllNamespaces, "all-namespaces", "", false, "List stacks among all Kubernetes namespaces")
+	flags.SetAnnotation("all-namespaces", "kubernetes", nil)
 	return cmd
 }
+
+func runList(cmd *cobra.Command, dockerCli command.Cli, opts options.List) error {
+	var err error
+	stacks := []*formatter.Stack{}
+	if dockerCli.ClientInfo().HasSwarm() {
+		ss, err := swarm.GetStacks(dockerCli)
+		if err != nil {
+			return err
+		}
+		stacks = append(stacks, ss...)
+	}
+	if dockerCli.ClientInfo().HasKubernetes() {
+		flags := cmd.Flags()
+		if dockerCli.ClientInfo().HasAll() && !flags.Changed("namespace") {
+			opts.AllNamespaces = true
+		}
+		if opts.AllNamespaces {
+			stacks, err = getStacksWithAllNamespaces(stacks, dockerCli, opts, flags)
+			if err != nil {
+				return err
+			}
+		} else {
+			stacks, err = getStacksWithNamespaces(stacks, dockerCli, opts, flags)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return format(dockerCli, opts, stacks)
+}
+
+func getStacksWithAllNamespaces(stacks []*formatter.Stack, dockerCli command.Cli, opts options.List, flags *pflag.FlagSet) ([]*formatter.Stack, error) {
+	var err error
+	stacks, err = getStacks(stacks, dockerCli, opts, kubernetes.NewOptions(flags))
+	if err == nil || !errors.IsForbidden(err) {
+		return stacks, err
+	}
+	nms, err := getUserVisibleNamespaces(dockerCli, err)
+	if err != nil {
+		return nil, err
+	}
+	opts.AllNamespaces = false
+	for _, nm := range nms.Items {
+		stacks, err = getStacks(stacks, dockerCli, opts, kubernetes.NewOptions(flags, nm.Name))
+		if err != nil {
+			return nil, err
+		}
+	}
+	return stacks, nil
+}
+
+func getStacksWithNamespaces(stacks []*formatter.Stack, dockerCli command.Cli, opts options.List, flags *pflag.FlagSet) ([]*formatter.Stack, error) {
+	mnms, err := getNamespaces(flags)
+	if err != nil {
+		return nil, err
+	}
+	for nm := range mnms {
+		stacks, err = getStacks(stacks, dockerCli, opts, kubernetes.NewOptions(flags, nm))
+		if err != nil {
+			return nil, err
+		}
+	}
+	return stacks, nil
+}
+
+func getStacks(stacks []*formatter.Stack, dockerCli command.Cli, opts options.List, kopts kubernetes.Options) ([]*formatter.Stack, error) {
+	kli, err := kubernetes.WrapCli(dockerCli, kopts)
+	if err != nil {
+		return nil, err
+	}
+	ss, err := kubernetes.GetStacks(kli, opts)
+	if err != nil {
+		return nil, err
+	}
+	return append(stacks, ss...), nil
+}
+
+func getNamespaces(flags *pflag.FlagSet) (map[string]struct{}, error) {
+	nms, err := flags.GetStringSlice("namespace")
+	if err != nil {
+		return nil, err
+	}
+	mnms := map[string]struct{}{}
+	for _, nm := range nms {
+		mnms[nm] = struct{}{}
+	}
+	return mnms, nil
+}
+
+func getUserVisibleNamespaces(dockerCli command.Cli, forbiddenErr error) (*core_v1.NamespaceList, error) {
```

Comment:
This is *really* hacky… I feel that we should extract this somewhere in the `kubernetes` package and make and make the hack there (a func or list of func, …) … This would allow, in the future, to code different methods to get the list of namespace.

This also duplicates a bunch of what the `cli`/`client` does… We may want to Update `https://github.com/moby/moby/blob/master/client/request.go` to make it a bit more composable — like allowing to define new endpoint to query (with a function that would handle the http response).